File: PY-SDK/admin/helpers/create_market_tests.py
from solana.keypair import Keypair
from pyaver.aver_client import AverClient
from solana.rpc.commitment import Confirmed
from pyaver.market import AverMarket
from asyncio import gather
from .init_market_instruction import InitMarketAccounts, InitMarketArgs
from .init_market_instruction import init_market
import time
from solana.publickey import PublicKey
from .supplement_init_market import SupplementInitMarketAccounts, supplement_init_market, SupplementInitMarketArgs

async def create_init_market_smoke_tests(
    client: AverClient,
    owner: Keypair,
    outcome_length: int,
    market: Keypair,
    market_authority: Keypair,
    program_id: PublicKey
):
    #CREATE MARKET
    init_market_args = InitMarketArgs(
        active_immediately=True,
        cranker_reward=0,
        fee_tier_collection_bps_rates=[0,0,0,0,0,0,0],
        going_in_play_flag=True,
        inplay_start_time=1682447605,
        market_name="test_market",
        max_quote_tokens_in=10000000000000000,
        max_quote_tokens_in_permission_capped=10000000000000000,
        min_new_order_base_size=1,
        min_new_order_quote_size=1,
        min_orderbook_base_size=1,
        number_of_outcomes=outcome_length,
        number_of_winners=1,
        permissioned_market_flag=False, 
        trading_cease_time=1682447605,
    )

    init_market_accs = InitMarketAccounts(
        market=market,
        market_authority=market_authority,
        payer=owner
    )

    program = await client.get_program_from_program_id(program_id)

    sig = await init_market(client, program, init_market_args, init_market_accs)
    con = await client.provider.connection.confirm_transaction(sig, Confirmed)
    assert(con['result']['value'][0]['err'] is None), 'Init market'
    
    #CREATE SUPP MARKET
    number_of_outcomes = outcome_length
    number_of_iterations = 1 if number_of_outcomes == 2 else number_of_outcomes
    
    coroutines = []

    for i in range(number_of_iterations):
        outcome_names = [f'outcome_{i}'] if number_of_outcomes > 2 else ['outcome_1', 'outcome_2']
        supplement_init_market_args = SupplementInitMarketArgs(
            event_capacity=100,
            nodes_capacity=100,
            outcome_id=i,
            outcome_names=outcome_names
        )

        supplement_init_market_accs = SupplementInitMarketAccounts(
            market=market.public_key,
            market_authority=market_authority,
            payer=owner
        )

        print(f'creating supplement init market: {i}')
        coroutines.append(supplement_init_market(client.program, supplement_init_market_args, supplement_init_market_accs))
    sigs = await gather(*coroutines)
    time.sleep(10)
    cons = await gather(*[client.provider.connection.confirm_transaction(sig, Confirmed) for sig in sigs])
    for con in cons:
        assert (con['result']['value'][0]['err'] is None), 'Supp market'
    time.sleep(10)

    #LOAD CREATED MARKET
    market_object = await AverMarket.load(client, market.public_key)
    [expected_market_store_pubkey, _] = AverMarket.derive_market_store_pubkey_and_bump(market.public_key)
    assert(market_object.market_pubkey == market.public_key), 'Loaded market matching'
    assert(market_object.market_state.market_authority == market_authority.public_key), 'Loaded matching market authority'
    assert(market_object.market_state.market_store == expected_market_store_pubkey), 'Loaded market store'
    assert(len(market_object.orderbooks) == market_object.market_state.number_of_outcomes), 'Outcome numbers'
    for o in market_object.orderbooks:
        assert(o.slab_asks is not None), 'Slab Asks'
        assert(o.slab_bids is not None), 'Slab Bids'

File: TS-SDK/admin/archive/__tests__/integration/utils/create-market-tests.ts
import { getOrCreateAssociatedTokenAccount } from '@solana/spl-token'
import { Keypair, Connection, PublicKey, SystemProgram } from '@solana/web3.js'
import BN from 'bn.js'
import { TransactionInstruction } from '@solana/web3.js/lib/index.esm'
import {
  makeInitMarketInstruction,
  makeSupplementInitMarketInstruction,
  SupplementInitMarketArgs,
} from './admin-instructions/init-market'
import { createAverClient } from './client'
import { AverClient, AVER_PROGRAM_ID, Market, signAndSendTransactionInstructions, USDC_DEVNET } from 'aver-ts'

jest.setTimeout(1000000)

export const getInitMarketSmokeTests = (
  owner: Keypair,
  outcomeLength: number,
  market: Keypair,
  marketAuthority: Keypair
) => {
  let client: AverClient
  let conn: Connection
  let keypair: Keypair

  beforeAll(async () => {
    client = (await createAverClient(owner)) as AverClient
    keypair = owner
    conn = client.connection
  })

  describe('init market', () => {
    let marketStore: PublicKey
    // TODO fails if too many outcomes as it requires more than 2 SOL
    let numberOfOutcomes: number = outcomeLength

    test('create a market successfully', async () => {
      const oracle = new Keypair()

      const [marketStorePubkey, marketStoreBump] = await Market.deriveMarketStorePubkeyAndBump(
        market.publicKey
      )
      marketStore = marketStorePubkey

      const [vaultAuthority, vaultBump] = await PublicKey.findProgramAddress(
        [market.publicKey.toBuffer()],
        AVER_PROGRAM_ID
      )
      const vaultQuoteAccount = (
        await getOrCreateAssociatedTokenAccount(conn, keypair, USDC_DEVNET, vaultAuthority, true)
      ).address

      const args = {
        numberOfOutcomes: numberOfOutcomes,
        numberOfWinners: 1,
        vaultBump: vaultBump,
        marketStoreBump: marketStoreBump,
        permissionedMarketFlag: false,
        minOrderbookBaseSize: new BN(1),
        minNewOrderBaseSize: new BN(1),
        minNewOrderQuoteSize: new BN(1),
        maxQuoteTokensIn: new BN(100000000000000),
        maxQuoteTokensInPermissionCapped: new BN(100000000000000),
        crankerReward: new BN(0),
        feeTierCollectionBpsRates: [
          new BN(20),
          new BN(18),
          new BN(16),
          new BN(14),
          new BN(12),
          new BN(10),
          new BN(0),
        ],
        marketName: 'Test market',
        goingInPlayFlag: true,
        activeImmediately: true,
        tradingCeaseTime: new BN(1682447605),
        inplayStartTime: new BN(1682447605),
      }
      const accs = {
        payer: keypair.publicKey,
        market: market.publicKey,
        marketAuthority: marketAuthority.publicKey,
        marketStore: marketStore,
        quoteTokenMint: USDC_DEVNET,
        vaultAuthority: vaultAuthority,
        quoteVault: vaultQuoteAccount,
        oracleFeed: SystemProgram.programId,
      }
      const initMarketIx = makeInitMarketInstruction(client, args, accs)

      const sig = await signAndSendTransactionInstructions(
        conn,
        [market, marketAuthority],
        keypair,
        [initMarketIx]
      )

      const confirmedSig = await conn.confirmTransaction(sig)

      expect(confirmedSig.value.err).toBeFalsy()
    })

    test('create supplement market correctly', async () => {
      // create the instructions
      const numOfAobs = numberOfOutcomes === 2 ? 1 : numberOfOutcomes
      const outcomeNames = Array(numberOfOutcomes)
        .fill(0)
        .map((v) => `outcomeNameBla`)
      const supplementIxs: TransactionInstruction[] = []

      for (let i = 0; i < numOfAobs; i++) {
        const outcomeName = numberOfOutcomes === 2 ? outcomeNames : [outcomeNames[i]]

        const args = {
          eventCapacity: 100,
          nodesCapacity: 100,
          outcomeId: i,
          outcomeNames: outcomeName,
        } as SupplementInitMarketArgs
        const accs = {
          market: market.publicKey,
          marketAuthority: marketAuthority.publicKey,
          marketStore: marketStore,
          payer: keypair.publicKey,
        }
        const initSupplementMarketIx = await makeSupplementInitMarketInstruction(client, args, accs)

        supplementIxs.push(initSupplementMarketIx)
      }

      let aobsToInit = supplementIxs

      do {
        const sigs = await Promise.all(
          aobsToInit.map((ix) =>
            signAndSendTransactionInstructions(conn, [marketAuthority], keypair, [ix])
          )
        )

        const aobSigs = await Promise.all(sigs.map((sig) => conn.confirmTransaction(sig)))

        aobsToInit = aobsToInit.filter((_aob, i) => !!aobSigs[i].value.err)
      } while (aobsToInit.length > 0)
    })

    test('check if market has been created correctly', async () => {
      const marketObject = await Market.load(client, market.publicKey)

      const [expectedMarketStorePubkey, _] = await Market.deriveMarketStorePubkeyAndBump(
        marketObject.pubkey
      )

      expect(marketObject.pubkey.toBase58()).toBe(market.publicKey.toBase58())
      expect(marketObject.marketAuthority.toBase58()).toBe(marketAuthority.publicKey.toBase58())
      expect(marketObject.marketStore.toBase58()).toBe(expectedMarketStorePubkey.toBase58())
      expect(marketObject?.orderbooks).toHaveLength(marketObject.numberOfOutcomes)
      marketObject?.orderbooks?.map((oo) => {
        expect(oo.slabAsks).toBeTruthy()
        expect(oo.slabBids).toBeTruthy()
      })
    })
  })
}
